### DISCUSSION TITLE
Dropping Python 2 support

### DISCUSSION THREAD

**Discussion by flit** [pyocd/pyOCD#1055] [General]
This discussion is about dropping of Python 2 support from pyOCD. Python 2 support is already deprecated, but there isn't a declared time when it will be fully dropped.

By no longer having to support Python 2, pyOCD can start taking advantage of recent Python 3 enhancements such as type hints for functions.

Does dropping Python 2 support actually affect anyone?
How soon should it be done?
Can pyOCD move to Python 3.7 and above (it's currently 3.6+)?

**Comment by cederom** [comment]
Python Foundation officially set python 2 sunset to 2015 then it was shifted to January 2020 [1]. FreeBSD Ports drops python 2 at the end of 2020 [2]. Everyone had enough time to switch to Python 3.

As a big fan of compatibility and adhering to standards I would say setting up Python 3 virtual environment and all required packages with pip is a matter of few seconds while keeping software compatible with both Python 2 and 3 which are not compatible can be painful and generates additional costly work. Because Python 2 is not supported for over a year now by its authors, initially scheduled in 2008 for EOL at 2015, I see no reason to support it with pyOCD anymore. We are in 2021. There has been 9 minor releases of Python 3 (most recent is 3.9.1) already so its a solid ground to move along. Switching host from Python 2 to 3 is not really painful they can even co-exist as independent virtual environments if someone really needs to stick to Python 2 for any reason.

Summing up I would drop Python 2 as soon as possible. People had over 5 years to update their code base :-)

[1] https://www.python.org/doc/sunset-python-2/
[2] https://www.freebsd.org/news/status/report-2020-10-2020-12.html#Python-2.7-removal-from-Ports

**Comment by flit** [comment]
Great argument, thanks. Fwiw, I haven't heard _any_ feedback against dropping Python 2. So I'm feeling pretty confident about dropping Python 2 in 0.30 (next feature release).

(I didn't realise Python 2 was originally sunset on 2015!)

**Reply by cederom** [reply]
I stick to FreeBSD which is very conservative when it comes to bleeding edge changes and any sort of fancy stuff, because it has maintenance and self-compatibility as its core values (as opposed to Linux), so people betting their money on BSD production systems just want them to work and not to change too often. When even FreeBSD drops Python 2 it really means its time to do it :-) :-) :-)

I had some doubts regarding such switch in 2015 or even 2017 but I am using only Python 3 virtual environments for a long time with no problem at all. I may not understand why minor numbers are changing so quickly, but that caused no problems for me so far. Python 3 is really stable and easy to use environment to work on :-)

ps: I would not switch directly to the latest minor though when it comes to security - I saw fuzzers discovering network and database related bugs on these - and I suspect more and more of them will show up on Python 2 :-)

**Reply by cederom** [reply]
new reply mode on github lol

**Reply by cederom** [reply]
well because there is a new "discussion" tab.. but they removed "releases" tab.. sweet brave new world ;-)

**Reply by flit** [reply]
Yeah, I haven't directly used Python 2 except for testing in a long time. We still have some internal Arm research environment scripts that require Python 2, but I'm in the process of converting those over right now.

**Comment by flit** [comment]
Fyi, I am planning to drop Python 2 in the next release, 0.30.0. The PR will be coming in probably tomorrow.

**Comment by flit** [comment]
It is done 😈: 8680ee61dcbec1f49081d17e8923526e20b27d29

**Comment by flit** [comment]
I'm testing now and fixing some bugs, and will release v0.30.0 as soon as possible.
